### PR TITLE
client: always refresh mds feature bits on session open

### PR DESCRIPTION
**Pull request by vshankar**
Fixes: http://tracker.ceph.com/issues/63188

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by mchangir — COMMENTED** [review]
could you rephrase commit title to:
**client: always refresh mds feature bits on session open**

**Comment by vshankar** [comment]
> could you rephrase commit title to: **client: always refresh mds feature bits on session open**

done.

File: src/client/Client.cc [hunk 1]
```diff
@@ -2402,6 +2408,7 @@ void Client::handle_client_session(const MConstRef<MClientSession>& m)
       if (session->state == MetaSession::STATE_OPEN) {
         ldout(cct, 10) << "mds." << from << " already opened, ignore it"
                        << dendl;
+	reinit_mds_features(session.get(), m);
```
Comment (dparmar18) [inline src/client/Client.cc:2411]: similar to my comment https://tracker.ceph.com/issues/63188#note-6 by is it possible to first check the feature bits held by the client vs what the mds holds currently? if not similar then reinit else dont.

Comment (vshankar) [reply src/client/Client.cc:2411]: Even if the features match, it doesn't hurt to just refresh those, isn't it? (compare+assign vs assign)

**Comment by dparmar18** [comment]
in a cluster with multiple active MDSs, is it possible that different MDSs are on different versions?

File: src/client/Client.cc (hunk 1 above)
Comment (dparmar18) [reply src/client/Client.cc:2411]: okay so there are no drawbacks involved

**Comment by vshankar** [comment]
> in a cluster with multiple active MDSs, is it possible that different MDSs are on different versions?

If you mean different ceph versions, then yes.

**Comment by dparmar18** [comment]
> > in a cluster with multiple active MDSs, is it possible that different MDSs are on different versions?
> 
> If you mean different ceph versions, then yes.

and the client must be maintaining some table (or some sort of data structure) to keep records of the MDSs and their feature bits right? So that client doesn't get baffled with what and how to communicate with mds.a, mds.b and mds.c if all three are on different ceph versions with possibly different set of feature bits to offer.

**Comment by vshankar** [comment]
> > > in a cluster with multiple active MDSs, is it possible that different MDSs are on different versions?
> > 
> > 
> > If you mean different ceph versions, then yes.
> 
> and the client must be maintaining some table (or some sort of data structure) to keep records of the MDSs and their feature bits right? So that client doesn't get baffled with what and how to communicate with mds.a, mds.b and mds.c if all three are on different ceph versions with possibly different set of feature bits to offer.

yes - those are maintained per (mds) session.

**Comment by dparmar18** [comment]
Can we test this out apart from the upgrade tests? Maybe add a dummy feature bit to the MDS and expect client to not crash?

**Comment by vshankar** [comment]
> Can we test this out apart from the upgrade tests? Maybe add a dummy feature bit to the MDS and expect client to not crash?

This yaml

```
fs/upgrade/featureful_client/old_client/{bluestore-bitmap centos_latest clusters/1-mds-2-client-micro conf/{client mds mon osd} overrides/{ignorelist_health ignorelist_wrongly_marked_down multimds/no pg-warn} tasks/{0-octopus 1-client 2-upgrade 3-compat_client/pacific}}
```

always reproduces this, so I think we can depend on that.

**Review by dparmar18 — APPROVED** [review]

File: src/client/Client.cc (hunk 1 above)
Comment (batrick) [inline src/client/Client.cc:2411]: Can we add a comment why this is necessary?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Client.cc (hunk 1 above)
Comment (vshankar) [reply src/client/Client.cc:2411]: Sure!

Comment (vshankar) [reply src/client/Client.cc:2417]: done.

**Comment by vshankar** [comment]
@batrick @dparmar18 fixed and updated. please check.

**Review by mchangir — APPROVED** [review]
LGTM

**Review by dparmar18 — APPROVED** [review]

**Review by kotreshhr — APPROVED** [review]

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2023-10-18_10:34:52-fs-wip-vshankar-testing-20231018.065603-testing-default-smithi/

**Comment by vshankar** [comment]
jenkins test make check

**Comment by dparmar18** [comment]
> https://pulpito.ceph.com/vshankar-2023-10-18_10:34:52-fs-wip-vshankar-testing-20231018.065603-testing-default-smithi/

i hope this includes https://github.com/ceph/ceph/pull/53734 which found out this issue

**Comment by vshankar** [comment]
> > https://pulpito.ceph.com/vshankar-2023-10-18_10:34:52-fs-wip-vshankar-testing-20231018.065603-testing-default-smithi/
> 
> i hope this includes #53734 which found out this issue

The issue was already seen in https://github.com/ceph/ceph/pull/50730 but I got to it rather late esp. when I looked at the core in your run, I figured that its the same issue.

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#18-Oct-2023

**Comment by vshankar** [comment]
rebase and updated for jenkins tests...

**Comment by vshankar** [comment]
jenkins test api
